### ISSUE TITLE
Lost Camera Rotation Anchor

### ISSUE THREAD

**Issue by Richard-L** [OpenRCT2/OpenRCT2#10873]
> OpenRCT2, v0.2.4-630-g425b05d01 (425b05d on develop) provided by GitHub
> Windows (x86-64)
Win7

Not really a bug, and probably not really OpenRCT2's "fault", but would there be any intelligent way to help with the rotation anchor point of the camera on parks with a lot of verticality?

https://www.nedesigns.com/park/526/the-island/
^Follow the "Giga Coaster" train as it reaches the top and then turns, and then rotate the view. You'll be looking at water elsewhere.

More examples:
- https://www.nedesigns.com/park/4398/karst/
- https://www.nedesigns.com/park/3639/via-lactea/

**Comment by ocalhoun6** [comment]
How can the game possibly know if you're looking at the coaster or the water behind it when the time comes to rotate the view?

Maybe if you're in the middle of building something, it could make sure the in-progress track piece remains centered in the view...

Not sure how it could possibly work when you're just going around and looking at things, though.

**Comment by Richard-L** [comment]
I agree with you it's hard if not impossible to somehow address this, but I
love your building mode idea. In fact would be necessary to test this
whilst in build mode, as perhaps this already sets a better focus?

Generally speaking though, how is the rotation anchor point determined in
the game? Is it a specific tile on the map? I would assume perhaps it's
always just the middle and lowest?

Disregard the coaster in "The Island" park though, even if I position the
camera so that the large land spike is in the middle, rotating the camera
will throw off my view completely. As this only happens with parks that
have a lot of objects placed very high, I'm assuming they have some sort of
camera influence.

Is there any merit in conceiving a routine where the game analyses the
average amount of "high objects" in the current camera position, and takes
that into account when choosing the rotation anchor?

**Comment by pizza2004** [comment]
I'm sure the camera just has coordinates of where it is pointing and if you have something very tall then you have to move it "beyond" where that thing is to truly see it.  I wonder if it would be possible to check what is in the center of the screen when you hit the rotate button and then rotate the camera around that thing instead of where it's "supposed" to be pointing.

I imagine that might be hard because you'd have to worry about it being tall and so you couldn't just find the position of the object and set the camera to it, you'd need to know how far to offset it to maintain the view of what you're looking at.

**Comment by ocalhoun6** [comment]
> so you couldn't just find the position of the object and set the camera to it, you'd need to know how far to offset it to maintain the view of what you're looking at.

If you know what particular object you want it centered on, there's already code existing to center the view on a particular object -- it's used in ride/peep windows and also in the title screen.

The real trick, though, is having the game know what you want it to stay centered on.

**Comment by pizza2004** [comment]
The trouble is that for something like a rollercoaster it would simply find the middle.  I don't know if it would even work for a specific piece of the track or not, but it would probably need to be changed anyway since that's not normally something you can target.

**Comment by Richard-L** [comment]
I think it would be an improvement already if you would see roughly the
same you were looking at prior to view rotation. Right now it feels as if
in addition to the rotation, the camera is panned elsewhere too. I remember
even in RCT 1 this was a problem.

**Comment by Richard-L** [comment]
[qSJiH1VTOp.zip](https://github.com/OpenRCT2/OpenRCT2/files/4308924/qSJiH1VTOp.zip)
^Quick screen recording to illustrate the problem. I'd love to understand on what basis the view is turned like that.

Fullsize map of the park shown: https://www.nedesigns.com/park/526/map/133/the-island/

**Comment by ocalhoun6** [comment]
The game doesn't know whether you're looking at that coaster or the water behind it.

Let's look at a simplified example.
Here, we have a spiral slide on top of a high tower and a haunted house in the distance behind it:
![Forest Frontiers 2020-03-09 15-52-05](https://user-images.githubusercontent.com/26573303/76264126-12aba080-621e-11ea-8e2a-a2b2335f36ba.png)

If we're looking at the spiral slide, we'll want the view to be like this when we rotate it:
![Forest Frontiers 2020-03-09 15-52-11](https://user-images.githubusercontent.com/26573303/76264149-26570700-621e-11ea-91b0-c76884e863d1.png)

But if we're looking at the haunted house, we'll want the view to be like this when we rotate it:
![Forest Frontiers 2020-03-09 15-52-08](https://user-images.githubusercontent.com/26573303/76264167-34a52300-621e-11ea-8fd9-a21e4c1a6182.png)

The game has no way of knowing which of those rotated views we want to see. It rotates based on assuming you're looking at something around 0 elevation, which is an assumption that works fine _most_ of the time, but as you've seen, it can be problematic when you're trying to look at something very high up. The problem is, the game doesn't have any reliable way to know when you're looking at something high up. How can it know whether you want to look at the spiral slide or the haunted house?

**Comment by Richard-L** [comment]
Thanks for that. Are you sure though it's actually anchoring on a tile, and
zero elevation? Because regardless of any height, that should mean your
haunted house should stay in the middle of the screen, no? Whether the
middle tile is high or not shouldn't matter, but it does.

I would love to hear from someone who can see the code to tell us how this
actually functions.

In theory, is it possible for the game to analyse the "frontmost" object
sitting in the middle of the screen in the moment of rotation? It would
then help if the rotation pivot could be set there. I understand this is
likely very difficult and hit or miss in a lot of cases. Just hoping to
brainstorm.
